### PR TITLE
CLI: Disallow blockhash/fee-calc lookups when offline

### PR DESCRIPTION
**Pull request by t-nelson**
#### Problem

`get_blockhash_fee_calculator()` allows cluster lookups when offline

#### Summary of Changes

Return error
Add test
Fix existing tests

Parse-side prevention of this case in a follow up PR

### CODE REVIEW

File: cli/src/cli.rs [hunk 1]
```diff
@@ -739,8 +739,14 @@ pub fn get_blockhash_fee_calculator(
         } else {
             (blockhash, rpc_client.get_recent_blockhash()?.1)
         }
-    } else {
+    } else if !sign_only {
         rpc_client.get_recent_blockhash()?
+    } else {
+        return Err(CliError::BadParameter(format!(
```
Comment (jstarry) [inline cli/src/cli.rs]: How about using https://docs.rs/clap/2.33.0/clap/struct.Arg.html#method.requires ?

Comment (t-nelson) [reply cli/src/cli.rs]: #7982 😉 

It unfortunately won't prevent us from writing incorrect tests, as I've fixed up here. Maybe `panic!()` would be more appropriate?

Comment (jstarry) [reply cli/src/cli.rs]: Ha nice! I guess ideally it wouldn't be possible to write incorrect tests, something like:
```rust
enum SignMethod {
  Online,
  Offline(Hash, bool)
}
```

Comment (t-nelson) [reply cli/src/cli.rs]: Getting there!
https://github.com/solana-labs/solana/blob/c13ab9f14ec136f94e89778e456150f956c8c864/cli/src/cli.rs#L77-L82

Comment (jstarry) [reply cli/src/cli.rs]: @t-nelson cool, I'd prefer that approach over the current approach

Comment (t-nelson) [reply cli/src/cli.rs]: Fully reworked as per request 😉

Comment (jstarry) [reply cli/src/cli.rs]: Ha great! Looks like a step in the right direction!

File: cli/src/offline.rs [hunk 2]
```diff
@@ -22,6 +25,49 @@ pub const SIGNER_ARG: ArgConstant<'static> = ArgConstant {
     help: "Provid a public-key/signature pair for the transaction",
 };
 
+#[derive(Clone, Debug, PartialEq)]
+pub enum BlockhashSpec {
```
Comment (jstarry) [inline cli/src/offline.rs]: Seems like `BlockhashSpec` is pretty coupled to `sign_only` now. What if this was instead something like:

```rust
enum TxMode {
  SignOnly,
  Send(Option<Hash>),
}
```

Basically combining the `blockhash` and `sign_only` arguments together

File: cli/src/offline.rs [hunk 3]
```diff
@@ -22,6 +25,49 @@ pub const SIGNER_ARG: ArgConstant<'static> = ArgConstant {
...
+    Partial(Hash),
+    Undeclared,
+}
+
+impl BlockhashSpec {
+    pub fn new_from_matches(matches: &ArgMatches<'_>) -> Self {
+        let blockhash = value_of(matches, BLOCKHASH_ARG.name);
+        let sign_only = matches.is_present(SIGN_ONLY_ARG.name);
+        match blockhash {
+            Some(hash) if sign_only => Self::Full(hash, FeeCalculator::default()),
```
[the remark this answers is not in the document]
Comment (t-nelson) [reply cli/src/offline.rs]: Pending https://github.com/solana-labs/solana/issues/7967 😉

[the remark this answers is not in the document]
Comment (jstarry) [reply cli/src/offline.rs]: Ha you're always one step ahead. Maybe we can move ahead assuming fees won't change much as rob mentioned

File: cli/src/offline.rs [hunk 4]
```diff
@@ -22,6 +25,49 @@ pub const SIGNER_ARG: ArgConstant<'static> = ArgConstant {
     help: "Provid a public-key/signature pair for the transaction",
 };
 
+#[derive(Clone, Debug, PartialEq)]
+pub enum BlockhashSpec {
+    Full(Hash, FeeCalculator),
+    Partial(Hash),
```
[the remark this answers is not in the document]
Comment (t-nelson) [reply cli/src/offline.rs]: I've come to the opinion that the online/offline terminology we've been using is limiting.  Instead, I think discussing signers as *present* or *absent* is more accurate and flexible.  There are cases where multiple signing authorities are constructing a transaction, each of whom are individually online, but their keys are physically separated.  In this case I may provide you a hash _and_ ask you to pay the fee, which you will likely want to look in to.

More to your point, yes that was the intent of the naming, my mind and the thesaurus we both failing me yesterday.  On fresh sleep, something like this seems more self-documenting.
```
enum BlockhashQuery {
  None(Hash, FeeCalculator),
  FeeCalculator(Hash),
  All,
}
```

File: cli/src/offline.rs (hunk 2 above)
Comment (t-nelson) [reply cli/src/offline.rs]: I feel like this somewhat over-encodes for the current state of the logic, which I'd like to move towards untangling.  My intent with the above is purely to intermediate the `rpc_client.get_recent_blockhash()` calls with user-supplied data.

File: cli/src/offline.rs [hunk 5]
```diff
@@ -22,6 +25,49 @@ pub const SIGNER_ARG: ArgConstant<'static> = ArgConstant {
...
+    Undeclared,
+}
+
+impl BlockhashSpec {
+    pub fn new_from_matches(matches: &ArgMatches<'_>) -> Self {
+        let blockhash = value_of(matches, BLOCKHASH_ARG.name);
+        let sign_only = matches.is_present(SIGN_ONLY_ARG.name);
```
[the remark this answers is not in the document]
Comment (t-nelson) [reply cli/src/offline.rs]: Sure! I can break out the bottom portion here to a separate function

**Comment by t-nelson** [comment]
I'm going to go ahead and break out the last commit here, so I can get it merged and unblock a few of my other branches.  Will rebase everything up as a result

File: cli/src/offline.rs (hunk 2 above)
Comment (jstarry) [reply cli/src/offline.rs]: Ah I see, I've been pushing for encoding the expected execution paths. Seems that we have two outcomes for a lot of our cli commands now: an offline signed transaction and an online signed and sent transaction. Wouldn't that make the code more readable and tests easier to write and understand?

Comment (t-nelson) [reply cli/src/offline.rs]: I'd like to work towards any combination of present and absentee signers, their online-ness irrelevant, and anyone who happens to have all of the signatures for a transaction may submit it to the network

Comment (jstarry) [reply cli/src/offline.rs]: Ah cool, nice to have that context. Works for me then!

**Comment by t-nelson** [comment]
@jstarry I think we're good here now 🙏

**Comment by t-nelson** [comment]
Awesome! Thanks! And no worries! I've played the timezone game a time or two :wink:

**Pull request merged**